proposals/configurable_tooling.md

```diff
@@ -0,0 +1,111 @@
+
+### Implementation
+1. Create a base container that includes needed tools `ks, kubectl, python2.7, python3.6`
+1. Add a new library kubeflow/core/kubeless.libsonnet to kubeflow/core that is included in kubeflow/core/all.libsonnet
+1. Define a set of kubeless Functions that match the `kf` subcommands EG:
+```yaml
+apiVersion: kubeless.io/v1beta1
+kind: Function
+metadata:
+  name: create
+spec:
+  handler: subcommand.create
+  runtime: python3.6
+  function: |
+  import json
+  import yaml
+  from kubernetes import client, config
+  config.load_incluster_config()
+  api=client.ExtensionsV1beta1Api()
+  ...
+```
+   1. Kubeless Functions should use the kubernetes python API for kubectl operations and the `ks` executable for ksonnet operations or possibly the new ksonnet-lib. The preference would be to utilize an ksonnet API.
+   1. Insure Kubeless core Functions are installed in containers to optimize execution times (see kubeless [documentation](http://kubeless.io/docs/building-functions/))
+1. Each kubeless Function will require a client subcommand plugin that can be dynamically loaded using golang plugin as a module. This plugin (.so) would be downloaded to a `kf` plugins directory as part of `kf command add ...`. It should also be autogenerated based on the Function Spec shown.
+
+```go
+// Function object
+type Function struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec FunctionSpec
+}
+
+// FunctionSpec contains func specification
+type FunctionSpec struct {
+	Handler                 string                          
+	Function                string                          
+	FunctionContentType     string                          
+	Checksum                string                          
+	Runtime                 string                          
+	Timeout                 string                          
+	Deps                    string                          
+	Deployment              v1beta1.Deployment              
+	ServiceSpec             v1.ServiceSpec
+	HorizontalPodAutoscaler v2beta1.HorizontalPodAutoscaler
+}
+```
+
+
+## Alternatives Considered
+1. A static set of subcommands that prescribes all available actions a data scientist would do.
+1. Using a different client framework other than spf13/cobra. Given that kubectl, kubeless and ks also use this framework there is an opportunity to leverage their subcommands.
```

Comment:
I'd also add #52 approach as an alternative - instead of FaaS we have running API endpoint, but problem itself I think is similar.